### PR TITLE
Validate publications with json schema

### PR DESCRIPTION
**Pull request by rjzondervan**
This adds the functionality to validate the data-field in a publication with JSON-schema definitions built in the publication type.

If there are no validation errors the field `validation` remains empty, while it will return data if there are validation errors. These errors follow the following form:
```json
{
    "/": ["the required properties (date) are missing"],
    "/date": ["The data must match the 'date' format"]
}
```

Please note that if there is an error on root level (like a missing parameter), the validator will not progress. (That means that the object above can never happen this way, so that object is just for illustrative purposes).

### CODE REVIEW

**Review by bbrands02 — APPROVED** [review]

File: lib/Db/PublicationType.php [hunk 1]
```diff
@@ -122,4 +123,33 @@ public function jsonSerialize(): array
...
+	public function getSchema(IURLGenerator $urlGenerator): object
+	{
+		$schema = [];
+		$schema['$schema']  = 'https://json-schema.org/draft/2020-12/schema';
+		$schema['$id']      = $urlGenerator->getAbsoluteURL($urlGenerator->linkToRoute('opencatalogi.publication_types.show', ['id' => $this->getUuid()]));
+		$schema['type']     = 'object';
+		$schema['required'] = [];
+		$schema['properties'] = [];
+
+		foreach($this->getProperties() as $name => $property) {
```
Comment (WilcoLouwerse) [inline lib/Db/PublicationType.php:143]: foreach (

File: lib/Service/ObjectService.php [hunk 2]
```diff
@@ -300,6 +306,11 @@ public function saveObject(string $objectType, array $object, bool $updateVersio
 	{
 		// Get the appropriate mapper for the object type
 		$mapper = $this->getMapper($objectType);
+
+		if($objectType === 'publication') {
```
Comment (WilcoLouwerse) [inline lib/Service/ObjectService.php:310]: if (

File: lib/Service/ValidationService.php [hunk 3]
```diff
@@ -14,107 +19,42 @@
...
-		    $catalog = $this->getCatalog($catalog);
-        } catch (OCSNotFoundException $exception) {
-            throw new OCSNotFoundException(message: $exception->getMessage());
-        }
+		$publication['validation'] = [];
 
-		// Check if the given publicationType is present in the catalog
-		if (in_array(needle: $publicationType, haystack: $catalog['publicationType']) === false) {
-			throw new OCSBadRequestException(message: 'Given publicationType object not present in catalog');
+		if($result->hasError()) {
```
Comment (WilcoLouwerse) [inline lib/Service/ValidationService.php:55]: if (

File: lib/Service/ValidationService.php [hunk 4]
```diff
@@ -14,107 +19,42 @@
...
-	 * @return array      The JSON Serialised catalog.
+	 * @param Publication $publication The publication to validate.
+	 * @return Publication The validated publication.
 	 *
-	 * @throws OCSNotFoundException If the catalog is not found.
-	 * @throws \GuzzleHttp\Exception\GuzzleException
+	 * @throws \OCP\AppFramework\Db\DoesNotExistException
+	 * @throws \OCP\AppFramework\Db\MultipleObjectsReturnedException
+	 * @throws \Psr\Container\ContainerExceptionInterface
+	 * @throws \Psr\Container\NotFoundExceptionInterface
```
Comment (WilcoLouwerse) [inline lib/Service/ValidationService.php:39]: Use imports instead?

**Review by WilcoLouwerse — APPROVED** [review]

**Review by WilcoLouwerse — APPROVED** [review]

**Pull request merged**